### PR TITLE
overhaul cookie section

### PR DESCRIPTION
**Pull request by jensens**
Improves: Cookie Section

### CODE REVIEW

File: develop/plone/sessions/cookies.rst [hunk 1]
```diff
@@ -2,71 +2,73 @@
...
 
     self.request.cookies.get("cookie_name", "default_value_if_cookie_not_set")
 
 Setting cookies
 ===============
 
-See `HTTPResponse.setCookie() <https://github.com/zopefoundation/Zope/blob/master/src/ZPublisher/HTTPResponse.py#L241>`_.
+Setting cookies is done on the response.
+Each cookie name there can be a new or an existing one.
+When sending the request, the information is turned into a properly formatted `Set-Cookie` header.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: Use double-backticks for inline literals in reST. Single is for markdown.

```
``Set-Cookie``
```

File: develop/plone/sessions/cookies.rst [hunk 2]
```diff
@@ -2,71 +2,73 @@
...
+.. code-block:: python
 
-You might want to tune up or clean cookies after some other part of Plone code has set them.
-You can do this in :doc:`post-publication event handler </develop/plone/serving/http_request_and_response>`.
+    self.request.response.setCookie("cookie_name", "cookie_value")
 
-Example ``cleancookies.py`` (needs ZCML subscriber registration too)::
+A cookie may have `cookie-attributes <https://en.wikipedia.org/wiki/HTTP_cookie#Cookie_attributes>`_ set.
+Those are passed as keyword arguments to ``setCookie``.
+Turning a cookie into a non-session cookie requires a expires='date' keyword,
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]:
```
Turning a cookie into a non-session cookie requires an ``expires='date'`` keyword and value.
```

File: develop/plone/sessions/cookies.rst [hunk 3]
```diff
@@ -2,71 +2,73 @@
...
+limiting the cookie to a path is a path='/somepath' keyword to the `setCookie()` method.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]:
```
Limiting the cookie to a path requires a ``path='/somepath'`` keyword and value for the ``setCookie()`` method.
```

File: develop/plone/sessions/cookies.rst [hunk 4]
```diff
@@ -2,71 +2,73 @@
...
-    Handling session and other cookies in Plone
+    Handling cookies in Plone
 
 
 Introduction
 ============
 
-Setting and getting cookies
+Cookies in Plone are read from the request and set on the response.
+There is an API for cookie handling.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: Where is the API? Got link, docstrings, or Sphinx autodoc?

File: develop/plone/sessions/cookies.rst [hunk 5]
```diff
@@ -2,71 +2,73 @@
...
 
-    from zope.interface import Interface
-    from zope.component import adapter
-    from plone.postpublicationhook.interfaces import IAfterPublicationEvent
+Clearing cookies
+================
 
+If a cookie needs to be removed, the browser has to be told to expire it.
+The ``expireCookie`` method does this.
+It always set ``max_age`` to ``0`` and the ``expires`` date to the past.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: s/set/sets

File: develop/plone/sessions/cookies.rst [hunk 6]
```diff
@@ -2,71 +2,73 @@
...
+Additional the cookie value will be set to ``deleted``.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: s/Additional/Additionally

File: develop/plone/sessions/cookies.rst [hunk 7]
```diff
@@ -2,71 +2,73 @@
...
+    self.request.response.expireCookie("cookie_name")
 
-        # All non-HTML payloads
-        if not request.response.headers["content-type"].startswith("text/html"):
-            # Rip-off I18N_language cookie
-            if "I18N_LANGUAGE" in request.response.cookies:
-                print "Cleaned up cookie for %s" % request["URL"]
-                del request.response.cookies["I18N_LANGUAGE"]
+The method allows additional attributes to be passed as kwargs, similar to ``setCookie``.
+Except ``max_age``, ``expires`` and ``value`` all is allowed, like `path` or other cookie attributes.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: I don't understand this sentence together with the previous. Can you clarify the intent? It seems to say that the following kwargs are reserved keywords, but it's not clear. Also use the Oxford comma and double backticks.

```
Except ``max_age``, ``expires``, and ``value`` all is allowed, like ``path`` or other cookie attributes.
```

File: develop/plone/sessions/cookies.rst [hunk 8]
```diff
@@ -105,8 +114,7 @@ Language cookie
 ---------------
 
 ``I18N_LANGUAGE`` is set by ``portal_languages`` tool.
-Disable it by *Use cookie for manual override* setting in
-``portal_languages``.
+Disable it by *Use cookie for manual override* setting in ``portal_languages``.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: Is that a GUI thing? If so, we have a directive for that:

```
by :guilabel:`Use cookie for manual override`
```

Also how does one disable it? Tick/untick a checkbox?

Also what does disabling the setting actually do and why would someone do it? This paragraph lacks an explanation for why to do this.

File: develop/plone/sessions/cookies.rst [hunk 9]
```diff
@@ -124,24 +132,67 @@ Setting session cookie lifetime
 Sanitizing cookies for the cache
 ================================
 
-You don't want to store HTTP responses with cookies in a front end cache
-server, because this would be a leak of other users' information.
+You don't want to store HTTP responses with cookies in a front end cache server,
+because this would be a leak of other users' information.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: - One sentence per line
- s/don't/do not

File: develop/plone/sessions/cookies.rst [hunk 10]
```diff
@@ -124,24 +132,67 @@ Setting session cookie lifetime
...
 
-Don't cache pages with cookies set. Also with multilingual sites it makes
-sense to have unique URLs for different translations as this greatly
-simplifies caching (you can ignore language cookie).
+Don't cache pages with cookies set.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: s/Don't/Do not

File: develop/plone/sessions/cookies.rst [hunk 11]
```diff
@@ -124,24 +132,67 @@ Setting session cookie lifetime
...
 
 :doc:`More info in Varnish section of this manual </manage/deploying/caching/varnish3>`.
 
+
+Late cleanup of HTTP response cookies
+=====================================
+
+You might want to tune up or clean cookies after some other part of Plone code has set them.
+You can do this after all processing is done and before the transaction is committed.
+Subscribing to `ZPublisher.interfaces.IPubBeforeCommit` event
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: Merge two lines into a sentence and use double backticks.

```
You can do this after all processing is done and before the transaction is committed by subscribing to the ``ZPublisher.interfaces.IPubBeforeCommit`` event.
```

File: develop/plone/sessions/cookies.rst [hunk 12]
```diff
@@ -124,24 +132,67 @@ Setting session cookie lifetime
...
+
+Put the code below in a ``cleancookies.py``.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: "in a file"

File: develop/plone/sessions/cookies.rst [hunk 13]
```diff
@@ -124,24 +132,67 @@ Setting session cookie lifetime
...
+
+.. code-block:: python
+
+    """Clean I18N cookies from non-HTML responses so that e.g. Image content,
+    which has language set, and is cross-linked across page,
+    don't inadvertiately change the language.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: s/inadvertiately/inadvertently

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Nice improvement! See requested changes for details.

File: develop/plone/sessions/cookies.rst (hunk 4 above)
Comment (jensens) [reply develop/plone/sessions/cookies.rst]: Ok, you're right, API is too much for 3 methods... I'll rephrase this.

File: develop/plone/sessions/cookies.rst (hunk 6 above)
Comment (stevepiercy) [reply develop/plone/sessions/cookies.rst]: Missed this one.

s/Additional/Additionally

File: develop/plone/sessions/cookies.rst (hunk 7 above)
Comment (stevepiercy) [reply develop/plone/sessions/cookies.rst]: These two sentences together on lines 70-71 still don't make sense to me. Can you clarify?

File: develop/plone/sessions/cookies.rst [hunk 14]
```diff
@@ -60,15 +60,15 @@ Clearing cookies
...
+It always sets ``max_age`` to ``0`` and the ``expires`` date to the past.
 Additional the cookie value will be set to ``deleted``.
 
 .. code-block:: python
 
     self.request.response.expireCookie("cookie_name")
 
 The method allows additional attributes to be passed as kwargs, similar to ``setCookie``.
-Except ``max_age``, ``expires`` and ``value`` all is allowed, like `path` or other cookie attributes.
+Except ``max_age``, ``expires`` and ``value`` all is allowed, like ``path`` or other cookie attributes.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]: These two sentences together on lines 70-71 still don't make sense to me. Can you clarify?

File: develop/plone/sessions/cookies.rst (hunk 8 above)
Comment (stevepiercy) [reply develop/plone/sessions/cookies.rst]: This one was missed.

**Review by stevepiercy — CHANGES_REQUESTED** [review]
You got most of them, but missed a few.

File: develop/plone/sessions/cookies.rst (hunk 14 above)
Comment (jensens) [reply develop/plone/sessions/cookies.rst]: You can set any additional attribute to the expiring cookie, except these 3.

File: develop/plone/sessions/cookies.rst [hunk 15]
```diff
@@ -2,71 +2,74 @@
...
-        #print "%s %s" % (request["URL"], request.response.cookies)
-
-        # All non-HTML payloads
-        if not request.response.headers["content-type"].startswith("text/html"):
-            # Rip-off I18N_language cookie
-            if "I18N_LANGUAGE" in request.response.cookies:
-                print "Cleaned up cookie for %s" % request["URL"]
-                del request.response.cookies["I18N_LANGUAGE"]
+The method ``expireCookie`` allows additional attributes to be passed as keyword arguments, similar to ``setCookie``.
+Above mentioned ``max_age``, ``expires`` and ``value`` are not allowed.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]:
```
The aforementioned keywords ``max_age``, ``expires``, and ``value`` are reserved and are not allowed.
```

File: develop/plone/sessions/cookies.rst [hunk 16]
```diff
@@ -2,71 +2,74 @@
...
+A common key and value pair used here is ``path``.
```
Comment (stevepiercy) [inline develop/plone/sessions/cookies.rst]:
```
A common keyword used here is ``path``.
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
There we go! Now it makes sense, and I have a couple of grammar fixes. Then it should be good to go. Thank you for your work on this! It's _so_ much better.

**Comment by jensens** [comment]
@stevepiercy thanks for your guidance.

**Review by stevepiercy — APPROVED** [review]

**Comment by stevepiercy** [comment]
Thank you!

**Pull request merged**